recirq/otoc/parallel_xeb.py

```diff
@@ -0,0 +1,725 @@
+
+    l_final = []  # type: List[Set[Tuple[cirq.GridQubit, cirq.GridQubit]]]
+    for gate_set in l_s:
+        if len(gate_set) != 0:
+            l_final.append(gate_set)
+
+    return l_final
+
+
+def build_xeb_circuits(qubits: Sequence[cirq.GridQubit],
+                       cycles: Sequence[int],
+                       benchmark_ops: Sequence[Union[
+                           cirq.Moment, Sequence[cirq.Moment]]] = None,
+                       random_seed: int = None,
+                       sq_rand_nums: Optional[np.ndarray] = None,
+                       reverse: bool = False,
+                       z_only: bool = False,
+                       ancilla: Optional[cirq.GridQubit] = None,
+                       cycles_per_echo: Optional[int] = None,
+                       light_cones: Optional[List[List[
+                           Set[cirq.GridQubit]]]] = None,
+                       echo_indices: Optional[np.ndarray] = None,
+                       ) -> Tuple[List[cirq.Circuit], np.ndarray]:
+    r"""Builds random circuits for cross entropy benchmarking (XEB).
+
+    A list of cirq.Circuits of varying lengths are generated, which are made
+    of random single-qubit gates and optional two-qubit gates.
+
+    Args:
+        qubits: The qubits to be involved in XEB.
+        cycles: The different numbers of cycles the random circuits will have.
+        benchmark_ops: The operations to be inserted between random
+            single-qubit gates. They can be one or more cirq.Moment objects,
+            or None (in which case no operation will be inserted between the
+            random single-qubit gates).
+        random_seed: The random seed for the single-qubit gates. If
+            unspecified, no random seed will be used.
+        sq_rand_nums: The random numbers representing the single-qubit gates.
+            They must be integers from 0 to 7 if the z_only is False,
+            and floats between -1 and 1 if z_only is True. The dimension of
+            sq_rand_nums should be len(qubits) by max(cycles). If
+            unspecified, the gates will be generated in-situ at random.
+        reverse: If True, benchmark_ops will be applied before the random
+            single-qubit gates in each cycle. Otherwise, it will be applied
+            after the random single-qubit gates.
+        z_only: Whether the single-qubit gates are to be random \pi/2
+            rotations around axes on the equatorial plane of the Bloch sphere
+            (z_only = False), or random rotations around the z-axis (z_only =
+            True). In the former case, the axes of rotations will be chosen
+            randomly from 8 evenly spaced axes ($\pi/4$, $\pi/2$ ... $7\pi/4$
+            radians from the x-axis). In the latter case, the angles of
+            rotation will be any random value between $-\pi$ and $\pi$.
+        ancilla: If specified, an additional qubit will be included in the
```

Comment:
is this actually useful? Is the ancilla included in the analysis/fitting?

Reply:
The ancilla is not really used in parallel XEB. It is added here for convenience, because the same circuit generation function is used in `build_otoc_circuits()`. This saves a lot of trouble in the OTOC circuit generation code.